### DISCUSSION TITLE
General approaches to the functionality for the app

### DISCUSSION THREAD

**Discussion by jdwilkin4** [jdwilkin4/contrast-ratio-repo#19] [Ideas]
This is the space to share ideas on how we are going to best approach implementing the functionality for the app.
Share your ideas and codepen examples here.
Don't worry about design or layout.  Just focus on the javascript. 
Then once we agree on the general approach, then the work will be broken up into tickets for people to work on.

**Comment by AndreyDmitriev39r** [comment]
Ok, I wrote some JS to convert rgb format to hex.
I leave it here for now:
https://codepen.io/AndreyDmitriev39r/pen/OJvEyqN

Is there a way to run it inside codepen like in some online-compiler?

**Reply by jdwilkin4** [reply]
I think is a good start. 👍 
If it looks like we are going to have a whole bunch of helper functions, we might consider moving them to a seperate folder like a `utils` folder. 
> Is there a way to run it inside codepen like in some online-compiler?

For now I think we can just use console.log statements. 

Later down the line, we can have conversations about using a testing framework like Jest to create unit tests.
But that discussion will happen after the functionality is implemented.

**Reply by jdwilkin4** [reply]
Another thing, we need to consider is how we are going to handle user inputs.
We want to make sure that the user is providing valid inputs.
We should start thinking about what messaging we want to provide the user if their input is invalid.

**Reply by Kernix13** [reply]
So it looks like you will be converting RGB values given by the user into HSL values. I haven't looked into the equations for contrast ratio, but is that how it is done? What if someone enters HSL values or an RGBA or HSLA? And what about 3-digit RGB values like #000 or #FFF?

**Reply by AndreyDmitriev39r** [reply]
> So it looks like you will be converting RGB values given by the user into HSL values. I haven't looked into the equations for contrast ratio, but is that how it is done? What if someone enters HSL values or an RGBA or HSLA? And what about 3-digit RGB values like #000 or #FFF?

I just wrote code for the specific type of conversion.
Of course we need to find formulas for each type of conversion we will use.
For this one I used simple algo from here:

https://www.developintelligence.com/blog/2017/02/rgb-to-hex-understanding-the-major-web-color-codes/#:~:text=First%20Value,code%20is%2012%2C%20or%20C

It was prety easy to google this stuff. I guess we can figure out approaches for defferent types of conversion

**Comment by jdwilkin4** [comment]
I created a couple of tickets for hex values and RGB values. 
We will have more tickets to tackle the other conversions.

**Comment by AndreyDmitriev39r** [comment]
Ok, this is interesting
Look at these 2 colors, guys
```
#123456
#DCEF34
```
If you will place it here:
https://contrast-ratio.com/#%23DCEF34-on-%23123456

they give you 9.95
However, precise contrast they give: 9.955418253263442

They are doing rounding with Math.floor or stuff like that I think.
I believe it's not correct.

In my work on issue 24 I am using toFixed(2)method for rounding
And for the same colors I am getting: 9.96
Precise value is the same.

I think my option is more accurate.
In general, I think, the user will not be bothered with such details.

Any thoughts, guys?

**Reply by Kernix13** [reply]
I would think Math.floor would be better because if you are going to have a +/- .01 variance, it should be towards a lower number. But then it's only .01 so I don't think it matters that much.
